buffer/src/test/java/io/netty/buffer/AbstractByteBufAllocatorTest.java

```diff
@@ -94,4 +95,52 @@ protected static void assertInstanceOf(ByteBuf buffer, Class<? extends ByteBuf>
         // Unwrap if needed
         assertTrue(clazz.isInstance(buffer instanceof SimpleLeakAwareByteBuf ? buffer.unwrap() : buffer));
     }
+
+    @Test
+    public void testUsedDirectMemory() {
+        T allocator = newAllocator(true);
+        Assume.assumeTrue(allocator instanceof ByteBufAllocatorMetric);
+        assertEquals(0, ((ByteBufAllocatorMetric) allocator).usedDirectMemory());
+        ByteBuf buffer = allocator.newDirectBuffer(1024, 4096);
+        int capacity = buffer.capacity();
+        assertEquals(expectedUsedMemory(allocator, capacity), ((ByteBufAllocatorMetric) allocator).usedDirectMemory());
+
+        // Double the size of the buffer
+        buffer.capacity(capacity << 1);
+        capacity = buffer.capacity();
+        assertEquals(buffer.toString(), expectedUsedMemory(allocator, capacity),
+                ((ByteBufAllocatorMetric) allocator).usedDirectMemory());
+
+        buffer.release();
+        assertEquals(expectedUsedMemoryAfterRelease(allocator, capacity),
+                ((ByteBufAllocatorMetric) allocator).usedDirectMemory());
+    }
+
+    @Test
+    public void testUsedHeapMemory() {
+        T allocator = newAllocator(true);
+        Assume.assumeTrue(allocator instanceof ByteBufAllocatorMetric);
```

Comment:
consider just doing the cast and the unit test will fail with a class cast exception and more information about types etc...